lib/config/wrapper.go

```diff
@@ -179,11 +206,77 @@ func (w *wrapper) RawCopy() Configuration {
 	return w.cfg.Copy()
 }
 
-// Replace swaps the current configuration object for the given one.
-func (w *wrapper) Replace(cfg Configuration) (Waiter, error) {
-	w.mut.Lock()
-	defer w.mut.Unlock()
-	return w.replaceLocked(cfg.Copy())
+func (w *wrapper) Modify(fn ModifyFunction) (Waiter, error) {
+	return w.modifyQueued(fn)
+}
+
+func (w *wrapper) modifyQueued(modifyFunc ModifyFunction) (Waiter, error) {
+	e := modifyEntry{
+		modifyFunc: modifyFunc,
+		res:        make(chan modifyResult),
+	}
+	select {
+	case w.queue <- e:
+	default:
+		return noopWaiter{}, errTooManyModifications
+	}
+	res := <-e.res
+	return res.w, res.err
+}
+
+func (w *wrapper) Serve(ctx context.Context) error {
+	defer w.serveSave()
+
+	var e modifyEntry
+	saveTimer := time.NewTimer(0)
+	<-saveTimer.C
+	saveTimerRunning := false
+	for {
+		select {
+		case e = <-w.queue:
+		case <-saveTimer.C:
+			w.serveSave()
+			saveTimerRunning = false
+			continue
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+
+		var waiter Waiter = noopWaiter{}
+		var err error
+
+		w.mut.Lock()
+		to := w.cfg.Copy()
+		if e.modifyFunc(&to) {
```

Comment:
As an alternative to returning a boolean we could serialize the config before and after the function call and compare to see if they changed something relevant or not. Then perhaps the modifier could/should return an error, for cases where it starts modifying the config but then realizes something went wrong and needs to bail out... but if that is not a thing that can happen currently we could leave the modifier without return value?

Reply:
It seems a bit ugly to me to serialize or reflect to see if the config changed, when the caller knows whether they changed the config or not. Not that it matters performance wise. Well that's my reasoning, I am not attached to either solution though :1234:

Reply:
I was thinking it a bit ugly that I had to say whether I modified the config or not, when the config is there so the caller can plainly see if I did or not. :) It also wasn't 100% certain what the bool return meant at first glance.

Since it's just for saving we could even just remember the last serialized version and when going to save check if it's the same and then skip it.

But if there are cases where we start changing the config and then return false because we change our mind then this is legitimate. (I didn't check if we ever do that, some of the auto accept and introducer stuff is convoluted.)

(I'm also not heavily invested in this, it's fine as is, really.)

Reply:
There's a bit of ugliness everywhere :)

> Since it's just for saving we could even just remember the last serialized version and when going to save check if it's the same and then skip it.

It's also to skip the subscriber verify and commit stages if unchanged.

> But if there are cases where we start changing the config and then return false because we change our mind then this is legitimate. (I didn't check if we ever do that, some of the auto accept and introducer stuff is convoluted.)

> It also wasn't 100% certain what the bool return meant at first glance.

That tips it in favor of not having a bool to me: Less potential for confusion.